### PR TITLE
Feature/update bll

### PR DESCRIPTION
**Pull request by kaallaash**
implement drawingService

### CODE REVIEW

File: DescriptiveGeometry/DG.API/Program.cs [hunk 1]
```diff
@@ -1,12 +1,18 @@
 using DG.BLL;
+using DG.DAL;
```
Comment (Blackolf931) [inline DescriptiveGeometry/DG.API/Program.cs:2]: ?

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 2]
```diff
@@ -1,37 +1,123 @@
...
     {
-        throw new NotImplementedException();
+        var drawingRow = await _dbContext.Drawings
+            .FirstOrDefaultAsync(d => d.Id == drawing.Id, cancellationToken);
+      
+        if (drawingRow is not null)
+        {
+            drawingRow.DrawingPhotoLink = drawing.DrawingPhotoLink;
+            drawingRow.DateUpdated = DateTimeOffset.Now;
+            await _dbContext.SaveChangesAsync(cancellationToken);
```
[the remark this answers is not in the document]
Comment (Blackolf931) [reply DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:91]: i think it should be on dal layer

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 3]
```diff
@@ -1,37 +1,123 @@
...
+    private async Task<List<Drawing>> ConvertToList(
+        IQueryable<DrawingRow> drawingRows,
+        CancellationToken cancellationToken)
     {
-        throw new NotImplementedException();
+        var drawingRowsList = await drawingRows.ToListAsync(cancellationToken);
+
+        var drawings = new List<Drawing>();
+
+        foreach (var drawingRow in drawingRowsList)
```
Comment (Blackolf931) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:116]: use auto mapper without foreach

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 4]
```diff
@@ -1,37 +1,123 @@
...
     }
 
-    public Task<Drawing> Get(int id, CancellationToken cancellationToken)
+    public async Task<List<Drawing>> GetAll(CancellationToken cancellationToken)
     {
-        throw new NotImplementedException();
+        var drawingRows = _dbContext.Drawings
+            .Include(d => d.Description);
+
+        return await ConvertToList(drawingRows, cancellationToken);
```
Comment (Blackolf931) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:70]: `ConvertToList` - ? why whe can't use .ToList() in the 68 line?

**Review by Blackolf931 — CHANGES_REQUESTED** [review]

File: DescriptiveGeometry/DG.BLL/AutoMapper/Profiles/DrawingRowToDrawingProfile.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        CreateMap<DrawingRow, Drawing>()
+            .ForMember(d => d.Description, opt =>
+                opt.MapFrom(dr => dr.Description.Text))
+            .ForMember(d => d.Points, opt =>
+                opt.MapFrom(dr => DeserializeToPoints(dr.Description.Points)))
+            .ForMember(d => d.DescriptionPhotoLink, opt =>
+                opt.MapFrom(dr => dr.Description.DescriptionPhotoLink));//DrawingPhotoLink
+    }
+
+    private static Dictionary<string, Coords> DeserializeToPoints(string str)
```
Comment (Blackolf931) [inline DescriptiveGeometry/DG.BLL/AutoMapper/Profiles/DrawingRowToDrawingProfile.cs:21]: ?

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 6]
```diff
@@ -1,37 +1,123 @@
...
 
-    public Task<Drawing> Update(Drawing drawing, CancellationToken cancellationToken)
+    private async Task<List<Drawing>> ConvertToList(
+        IQueryable<DrawingRow> drawingRows,
+        CancellationToken cancellationToken)
     {
-        throw new NotImplementedException();
+        var drawingRowsList = await drawingRows.ToListAsync(cancellationToken);
+
+        var drawings = new List<Drawing>();
```
Comment (stas1408123) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:114]: Return entity in this case

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 7]
```diff
@@ -1,37 +1,123 @@
...
+            descriptionDrawingRow.DescriptionPhotoLink = drawing.DescriptionPhotoLink;
+            await _dbContext.SaveChangesAsync(cancellationToken);
+        }
+
+        return drawing;
     }
 
-    public Task<Drawing> Update(Drawing drawing, CancellationToken cancellationToken)
+    private async Task<List<Drawing>> ConvertToList(
+        IQueryable<DrawingRow> drawingRows,
```
Comment (stas1408123) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:109]: Is it really needed?

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 8]
```diff
@@ -1,37 +1,123 @@
...
-        throw new NotImplementedException();
+        var drawingRows = _dbContext.Drawings
+            .Include(d => d.Description)
+            .Where(d => d.Description.Text == description);
+
+        return await ConvertToList(drawingRows, cancellationToken);
     }
 
-    public Task<List<Drawing>> GetByDescription(string query, CancellationToken cancellationToken)
+    public async Task<Drawing> Update(Drawing drawing, CancellationToken cancellationToken)
```
Comment (stas1408123) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:82]: https://tldrify.com/19fg

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 9]
```diff
@@ -1,37 +1,123 @@
...
+        return await ConvertToList(drawingRows, cancellationToken);
     }
 
-    public Task<List<Drawing>> GetAll(CancellationToken cancellationToken)
+    public async Task<List<Drawing>> GetByDescription(string description, CancellationToken cancellationToken)
     {
-        throw new NotImplementedException();
+        var drawingRows = _dbContext.Drawings
+            .Include(d => d.Description)
+            .Where(d => d.Description.Text == description);
```
Comment (stas1408123) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:77]: ToListAsync()

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 10]
```diff
@@ -1,37 +1,123 @@
...
+        drawingRow.DateCreated = DateTimeOffset.Now;
+        drawingRow.DateUpdated = DateTimeOffset.Now;
+        drawingRow.DownloadsCount = 0;
+
+        await _dbContext.Drawings.AddAsync(drawingRow, cancellationToken);
+
+        var drawingDescriptionRow = _mapper.Map<DrawingDescriptionRow>(drawing);
+        drawingDescriptionRow.Drawing = drawingRow;
+
+        await _dbContext.DrawingDescription.AddAsync(drawingDescriptionRow, cancellationToken);
```
Comment (stas1408123) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:36]: You will add a DrawingDescriptionRow if it is included in Drawings.

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs [hunk 11]
```diff
@@ -1,37 +1,123 @@
...
+        await _dbContext.SaveChangesAsync(cancellationToken);
+
+        drawing.Id = drawingRow.Id;
```
Comment (stas1408123) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:39]: This has not been applied to the DB

File: DescriptiveGeometry/DG.DAL/DependencyRigistrator.cs [hunk 12]
```diff
@@ -16,6 +16,7 @@ public static class DependencyRegistrar
         var optionsBuilder = new DbContextOptionsBuilder<DatabaseContext>();
         optionsBuilder.UseSqlServer(
             configuration.GetConnectionString("DescriptiveGeometryDb"));
+        services.AddDbContext<DatabaseContext>();
```
Comment (stas1408123) [inline DescriptiveGeometry/DG.DAL/DependencyRigistrator.cs:19]: https://learn.microsoft.com/ru-ru/dotnet/api/microsoft.extensions.dependencyinjection.entityframeworkservicecollectionextensions.adddbcontext?view=efcore-6.0

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs (hunk 7 above)
Comment (stas1408123) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:109]: Is it needed?

File: DescriptiveGeometry/DG.BLL/Services/DrawingService.cs (hunk 8 above)
Comment (stas1408123) [inline DescriptiveGeometry/DG.BLL/Services/DrawingService.cs:82]: DrawingEntity

**Review by stas1408123 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**